sqlx-postgres/src/types/geometry/box.rs

```diff
@@ -0,0 +1,257 @@
+/// Representation: `((x1,y1),(x2,y2))`
+///
+/// Boxes are represented by pairs of points that are opposite corners of the box. Values of type box are specified using any of the following syntaxes:
+///
+/// ```
+/// ( ( x1 , y1 ) , ( x2 , y2 ) )
+/// ( x1 , y1 ) , ( x2 , y2 )
+///   x1 , y1   ,   x2 , y2
+/// ```
+/// where `(x1,y1) and (x2,y2)` are any two opposite corners of the box.
+/// Boxes are output using the second syntax.
+/// Any two opposite corners can be supplied on input, but the values will be reordered as needed to store the upper right and lower left corners, in that order.
+///
+/// See https://www.postgresql.org/docs/16/datatype-geometric.html#DATATYPE-GEOMETRIC-BOXES
+#[derive(Debug, Clone, PartialEq)]
+pub struct PgBox {
+    pub x1: f64,
+    pub y1: f64,
+    pub x2: f64,
+    pub y2: f64,
+}
+
+impl Type<Postgres> for PgBox {
+    fn type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("box")
+    }
+}
+
+impl PgHasArrayType for PgBox {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("_box")
+    }
+}
+
+impl<'r> Decode<'r, Postgres> for PgBox {
+    fn decode(value: PgValueRef<'r>) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        match value.format() {
+            PgValueFormat::Text => Ok(PgBox::from_str(value.as_str()?)?),
+            PgValueFormat::Binary => Ok(PgBox::from_bytes(value.as_bytes()?)?),
+        }
+    }
+}
+
+impl<'q> Encode<'q, Postgres> for PgBox {
+    fn produces(&self) -> Option<PgTypeInfo> {
+        Some(PgTypeInfo::with_name("box"))
+    }
+
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> Result<IsNull, BoxDynError> {
+        self.serialize(buf)?;
+        Ok(IsNull::No)
+    }
+}
+
+impl FromStr for PgBox {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let sanitised = s.replace(&['(', ')', '[', ']', ' '][..], "");
+        let mut parts = sanitised.splitn(4, ",");
+
+        let x1 = parts
+            .next()
+            .and_then(|s| s.parse::<f64>().ok())
+            .ok_or(Error::Decode(
+                format!("{}: could not get x1 from {}", ERROR, s).into(),
+            ))?;
+
+        let y1 = parts
+            .next()
+            .and_then(|s| s.parse::<f64>().ok())
+            .ok_or(Error::Decode(
+                format!("{}: could not get y1 from {}", ERROR, s).into(),
+            ))?;
+
+        let x2 = parts
+            .next()
+            .and_then(|s| s.parse::<f64>().ok())
+            .ok_or(Error::Decode(
+                format!("{}: could not get x2 from {}", ERROR, s).into(),
+            ))?;
+
+        let y2 = parts
+            .next()
+            .and_then(|s| s.parse::<f64>().ok())
+            .ok_or(Error::Decode(
+                format!("{}: could not get y2 from {}", ERROR, s).into(),
+            ))?;
+
+        Ok(PgBox { x1, y1, x2, y2 })
+    }
+}
+
+impl PgBox {
+    fn from_bytes(mut bytes: &[u8]) -> Result<PgBox, Error> {
```

Comment:
PgCube uses a BoxDynError here, is that needed here?

https://github.com/jayy-lmao/sqlx/blob/feat/support-geometry-postgres/sqlx-postgres/src/types/cube.rs#L138

Reply:
Any error returned from `Decode::decode()` gets wrapped in [`sqlx::Error::Decode()`](https://docs.rs/sqlx/latest/sqlx/enum.Error.html#variant.Decode).

Thus, if you return `sqlx::Error` it'll turn into `sqlx::Error::Decode(Box::new(sqlx::Error::Decode()))` which is redundant.

`String` is directly convertible to `Box<dyn Error>` with `.into()` or the `?` operator, so you don't need to wrap it in anything.

It'd probably be better to switch to a real error type that captures more context, but that's orthogonal to this PR.

Reply:
Thank you for the explanation